src/lava/magma/core/model/py/ports.py

```diff
@@ -14,31 +14,80 @@
 
 
 class AbstractPyPort(AbstractPortImplementation):
+    """Abstract class for ports implemented in python. A port must have one or
+    multiple csp ports."""
     @property
     @abstractmethod
     def csp_ports(self) -> ty.List[AbstractCspPort]:
-        """Returns all csp ports of the port."""
+        """
+        Abstract property to get all csp ports of a port.
+
+        Returns
+        -------
+        A list of all csp ports used by the port.
+        """
         pass
 
 
-class AbstractIOPyPort(AbstractPyPort):
+class AbstractPyIOPort(AbstractPyPort):
+    """Abstract class of an input/output port implemented in python.
+
+    Parameters
+    ----------
+    csp_ports : list
+        A list of csp ports used by this IO port.
+
+    process_model : AbstractProcessModel
+        The process model used by the process of the port.
+
+    shape : tuple, default=tuple()
+        The shape of the port.
+
+    d_type: type, default=int
+        The data type of the port.
+
+    Attributes
+    ----------
+    _csp_ports : list
+        A list of csp ports used by this IO port.
+    """
     def __init__(self,
                  csp_ports: ty.List[AbstractCspPort],
                  process_model: AbstractProcessModel,
                  shape: ty.Tuple[int, ...] = tuple(),
                  d_type: type = int):
+
         self._csp_ports = csp_ports
         super().__init__(process_model, shape, d_type)
 
     @property
     def csp_ports(self) -> ty.List[AbstractCspPort]:
-        """Returns all csp ports of the port."""
+        """Property to get all csp ports used by the port.
+
+        Returns
+        -------
+        A list of all csp ports used by the port.
+        """
         return self._csp_ports
 
 
-class PyInPort(AbstractIOPyPort):
+class PyInPort(AbstractPyIOPort):
     """Python implementation of InPort used within AbstractPyProcessModel.
     If buffer is empty, recv() will be blocking.
+
+    Attributes
+    ----------
+    VEC_DENSE : PyInPortVectorDense, default=None
```

Comment:
Not really helpful explanation. This class attribute does not "specify" anything. It's just a class attribute that returns the type of one of four different sub types of PyInputPorts like PyINPortVectorDense, etc.

Then explain what this is even good for: Used to select type of InPorts via LavaPyType declarations in PyProcModels.

Reply:
The background explanation should probably pulled out into the main docstring such that in the the Attributes section you only explain what VectorDense, VectorSparse, ...

Also, be sure to say "Class attributes" so it's not confused with instance attributes that are usually defined in __init__ or as part of dataclasses.